client/ayon_core/hosts/zbrush/api/communication_server.py

```diff
@@ -0,0 +1,746 @@
+
+        os.environ["WEBSOCKET_URL"] = "ws://localhost:{}".format(
+            self.websocket_server.port
+        )
+
+        log.info("Added request handler for url: {}".format(
+            os.environ["WEBSOCKET_URL"]
+        ))
+
+        self._start_webserver()
+
+        # Start Zbrush when server is running
+        self._launch_zbrush(launch_args)
+
+        log.info("Waiting for client connection")
+        while True:
+            if self.process.poll() is not None:
+                log.debug("Host process is not alive. Exiting")
+                self._exit(1)
+                return
+
+            if self.websocket_rpc.client_connected():
+                log.info("Client has connected")
+                break
+            time.sleep(0.5)
+
+        self._on_client_connect()
+
+        emit_event("application.launched")
+
+    def _on_client_connect(self):
+        self._initial_textfile_write()
+
+    def _initial_textfile_write(self):
+        """Make sure the zscript menu already installed
+        when starting Zbrush."""
+        menu_txt_filepath = os.path.join(
+            ZBRUSH_HOST_DIR, "api", "zscripts", "ayon_zbrush_menu.txt")
+        menu_filepath = menu_txt_filepath.replace("\\", "/")
+        if menu_filepath is None:
+            log.warning(
+                "No ayon menu installed in Zbrush."
+            )
+
+    def _client(self):
+        if not self.websocket_rpc:
+            log.warning("Communicator's server did not start yet.")
+            return None
+
+        for client in self.websocket_rpc.clients:
+            if not client.ws.closed:
+                return client
+        log.warning("Client is not yet connected to Communicator.")
+        return None
+
+    def client(self):
+        if not self._connected_client or self._connected_client.ws.closed:
+            self._connected_client = self._client()
+        return self._connected_client
+
+    def send_request(self, method, params=None):
+        client = self.client()
+        if not client:
+            return
+
+        return self.websocket_rpc.send_request(
+            client, method, params
+        )
+
+    def send_notification(self, method, params=None):
+        client = self.client()
+        if not client:
+            return
+
+        self.websocket_rpc.send_notification(
+            client, method, params
+        )
+
+    def execute_zscript(self, zscript):
+        """Execute passed zscript in Zbrush."""
+        zbrush_exe = os.environ["ZBRUSH_EXE"]
+        output_file = tempfile.NamedTemporaryFile(
+            mode="w", prefix="a_zb_", suffix=".txt", delete=False
+        )
+        output_file.close()
+        output_filepath = output_file.name.replace("\\", "/")
+        with open(output_filepath, "wt") as tmp:
+            tmp.write(zscript)
+        subprocess.call([zbrush_exe, output_filepath],
+                        shell=True)
+        # remove the temp file
+        os.remove(output_filepath)
```

Comment:
Doesn't this work?
```suggestion
        with tempfile.NamedTemporaryFile(
            mode="w", prefix="a_zb_", suffix=".txt"
        ) as f:
            f.write(zscript)
            subprocess.call([zbrush_exe, f.name], shell=True)
```

Why do we need to close it first, then write to it, close it again, etc.

Reply:
> with tempfile.NamedTemporaryFile(
>             mode="wt", prefix="a_zb_", suffix=".txt"
>         ) as f:
>             f.write(zscript)
>             subprocess.call([zbrush_exe, f.name], shell=True)

it will error out when i tested with loading model, so no.

Reply:
Could you provide the error?

Reply:
> with tempfile.NamedTemporaryFile(
>             mode="wt", prefix="a_zb_", suffix=".txt"
>         ) as f:
>             f.write(zscript)
>             subprocess.call([zbrush_exe, f.name], shell=True)

![image](https://github.com/ynput/ayon-core/assets/64118225/68c717af-b6d9-4a98-a817-9c7646a5bbcd)